docs/rfcs/0561-list-metadata-reuse.md

```diff
@@ -0,0 +1,159 @@
+- Proposal Name: `list_metadata_reuse`
+- Start Date: 2022-08-23
+- RFC PR: [datafuselabs/opendal#561](https://github.com/datafuselabs/opendal/pull/561)
+- Tracking Issue: None
+
+# Summary
+
+Reuse metadata returned during listing, by defining a new `MetaLite` struct storing some metadata fields, and embedding it to `DirEntry`.
+
+# Motivation
+
+Users may expect to browse metadata of some directories' child files and directories. `list()` seems to be an ideal way to complete this job. 
+
+Thus, they start iterating on it, but soon they realized the `DirEntry`, which could only offer the name (or path, more precisely) and access mode of the object.
+
+So they have to call `metadata()` for each name they extracted from the iterator.
+
+The final example looks like:
+
+```rust
+let op = Operator::from_env(Scheme::Gcs)?.batch();
+
+// here is a network request
+let mut dir_stream = op.walk("/dir/to/walk")?;
+
+while let Some(Ok(file)) = dir_stream.next().await {
+    let path = file.path();
+
+    // here is another network request
+    let size = file.metadata().await?.content_length();
+    println!("size of file {} is {}B", path, size);
+}
+```
+
+But...wait! many storage-services returns object metadata when listing, like HDFS, AWS and GCS. The rust standard library returns metadata when listing local file systems, too.
+
+In the previous versions of OpenDAL we just simply ignored them. This wastes users' time on requesting on metadata.
+
+# Guide-level explanation
+
+The loop in main will be changed to the following code with this RFC:
+
+```rust
+while let Some(Ok(file)) = dir_stream.next().await {
+    let size = if let Some(len) = file.content_length() {
+        len
+    } else {
+        file.metadata().await?.content_length();
+    };
+    let name = file.path();
+    println!("size of file {} is {}B", path, size);
+}
+```
+
+# Reference-level explanation
+
+This RFC suggests embedding some metadata fields in `DirEntry`
+
+```rust
+struct MetaLite {
```

Comment:
Introducing a new struct makes implementation more complex.

I'm rethinking your ideas about embedding `ObjectMetadata` inside `DirEntry`.

Maybe we can add the following changes (internally, no API breakages in public API):

- `ObjectMetadata` `content_length` convert into `Option<u64>`
  - For `stat`, it's required to set `content_length` so that we can `expect` it in `metadata`'s API.
  - For `list`, this field is allowed to be `None`.
- Remove `DirEntry`'s `ObjectMode`, we can use the value from `ObjectMetadata`.
- Other fields are exactly the same.

After these, we can impl APIs for `DirEntry` directly.

Reply:
It's the second way I described before, in fact.

And adding it also comes along with an `ETag` field, I'm afraid it's not part of the current plan.

Adding a `MetaLite` struct is better, I think.

Reply:
S3 will return `ETag` too:

```xml
    <Contents>
        <Key>my-image.jpg</Key>
        <LastModified>2009-10-12T17:50:30.000Z</LastModified>
        <ETag>"fba9dede5f27731c9771645a39863328"</ETag>
        <Size>434234</Size>
        <StorageClass>STANDARD</StorageClass>
        <Owner>
            <ID>75aa57f09aa0c8caeab4f8c24e99d10f8e7faeebf76c078efc7c6caea54ba06a</ID>
            <DisplayName>mtd@amazon.com</DisplayName>
        </Owner>
    </Contents>
```

Reply:
IMHO, `MetaLite` will become large in the future. The only difference is that `MetaLite` doesn't have an `ObjectMode`. As a tradeoff, the service implementor needs to use a new struct instead of using `ObjectMeta` directly. I think `ObjectMode` doesn't worth that.

Reply:
> IMHO, `MetaLite` will become large in the future. The only difference is that `MetaLite` doesn't have an `ObjectMode`. As a tradeoff, the service implementor needs to use a new struct instead of using `ObjectMeta` directly. I think `ObjectMode` doesn't worth that.

I think the metadata `DirEntry` contains should be a subset of `ObjectMeta`. If we have some metadata fields that are not part of `ObjectMeta`, `ObjectMeta` should also be extended.

Reply:
> I think the metadata `DirEntry` contains should be a subset of `ObjectMeta`.

I agree with this conclusion.

There are three plans here:

- Plan A: Add fields in `DirEntry` directly.
- Plan B: Add a new struct `MetaLite`, which is the subset of `ObjectMetadata` (@ClSlaid prefer)
- Plan C: Use `ObjectMetadata` here directly. (@Xuanwo prefer)

Our public API has agreed, and the different plans are just internal implementation details. How about starting from Plan A? We can try them in the actual use cases and come back here (with new RFCs) to decide whether we pick Plan B or C.

If you agree with this idea, you can take `Plan A` as proposed and leave a section in `Future possibilities`. Otherwise, please feel free to comment here.

Reply:
Ok, let's get started.